adapters/smaato/smaato.go

```diff
@@ -0,0 +1,288 @@
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	uri := a.URI
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     uri,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+// MakeBids unpacks the server's response into Bids.
+func (a *SmaatoAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+
+			var markupError error
+			bid.AdM, markupError = renderAdMarkup(getAdMarkupType(response, bid.AdM), bid.AdM)
+			if markupError != nil {
+				fmt.Println(markupError)
+				continue // no bid when broken ad markup
+			}
+
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: openrtb_ext.BidTypeBanner,
+			})
+		}
+	}
+	return bidResponse, nil
+}
+
+func renderAdMarkup(adMarkupType adMarkupType, adMarkup string) (string, error) {
+	var markupError error
+	var adm string
+	switch adMarkupType {
+	case smtAdTypeImg:
+		adm, markupError = extractAdmImage(adMarkup)
+	case smtAdTypeRichmedia:
+		adm, markupError = extractAdmRichMedia(adMarkup)
+	default:
+		return "", fmt.Errorf("Unknown markup type %s", adMarkupType)
+	}
+	return adm, markupError
+}
+
+func getAdMarkupType(response *adapters.ResponseData, adMarkup string) adMarkupType {
+	if admType := adMarkupType(response.Headers.Get("X-SMT-ADTYPE")); admType != "" {
+		return admType
+	}
+	if strings.HasPrefix(adMarkup, `{"image":`) {
+		return smtAdTypeImg
+	}
+	if strings.HasPrefix(adMarkup, `{"richmedia":`) {
+		return smtAdTypeRichmedia
+	}
+	return ""
+}
+
+func assignBannerSize(banner *openrtb.Banner) (*openrtb.Banner, error) {
+	if banner.W != nil && banner.H != nil {
+		return banner, nil
+	}
+	if len(banner.Format) == 0 {
+		return banner, fmt.Errorf("No sizes provided for Banner %v", banner.Format)
+	}
+	bannerCopy := banner
```

Comment:
This is not actually making a copy. You need to deference the banner:
`bannerCopy := *banner`

Reply:
ahh got it, made the change